### PR TITLE
[`flake8-tidy-imports`] Add `TID253`

### PR DESCRIPTION
**Pull request by durumu**
## Summary

Add a new rule `TID253` (`banned-module-level-imports`), to ban a user-specified list of imports from appearing at module level. This rule doesn't exist in `flake8-tidy-imports`, so it's unique to Ruff. The implementation is pretty similar to `TID251`.

Briefly discussed [here](https://github.com/astral-sh/ruff/discussions/6370).

## Test Plan

Added a new test case, checking that inline imports are allowed and that non-inline imports from the banned list are disallowed.

### CODE REVIEW

File: crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+/// ## What it does
+/// Checks for banned imports at module level. The banned imports are allowed inline, such as
+/// within a function definition or an `if TYPE_CHECKING:` block.
+///
+/// ## Why is this bad?
+/// Some modules take a long time to import. Library authors might want to ensure that you only pay
+/// the import cost for these modules if you directly use them, rather than if you import a module
+/// that happens to use an expensive module in one of its functions.
+///
```
Comment (konstin) [inline crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs:17]: this needs an example

File: crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+enum NameMatchPolicy {
+    ExactOnly,
+    ExactOrParents,
+}
+
+fn banned_at_module_level_with_policy<T>(
+    checker: &mut Checker,
+    name: &str,
+    stmt: &Stmt,
+    located: &T,
```
Comment (konstin) [inline crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs:42]: nit: you can avoid the generic and pass the `TextRange` directly instead

**Review by konstin — APPROVED** [review]

**Comment by durumu** [comment]
Sorry for the ping @dhruvmanila, I don't think you need to review this one -- I accidentally included some extra commits when I pushed 😅 (first github PR!)

File: crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs (hunk 1 above)
Comment (durumu) [reply crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs:19]: Added!

File: crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs (hunk 2 above)
Comment (durumu) [reply crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs]: Good idea, thanks!

**Comment by charliermarsh** [comment]
Generally looks good, will try to get to this soon.

File: crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    checker: &mut Checker,
+    name: &str,
+    stmt: &Stmt,
+    text_range: TextRange,
+    locator: &Locator,
+    policy: &NameMatchPolicy,
+) {
+    if !locator.is_at_start_of_line(stmt.start()) {
+        return;
+    }
```
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs:50]: I think more robust here would be to use `checker.semantic().at_top_level()`, which will base its decision on scoping and parent statements. It's possible to have statements that aren't at the start of the line, but are at the top level, like:
```python
x = 1; import tensorflow
```

Comment (durumu) [reply crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs:50]: Thanks! I shamelessly copied this from the `module_import_not_at_top_of_file` rule (`E402`), so I think that will need to be fixed too.

Comment (charliermarsh) [reply crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs:89]: Thanks for pointing that out, good call! Fixed in https://github.com/astral-sh/ruff/pull/6526/files.

**Pull request merged**